你将学习处理文件，让程序能够快速地分析大量的数据；     
你将学习错误处理，避免程序在面对意外情形时崩溃；     
你将学习**异常**，它们是Python创建的特殊对象，用于管理程序运行时出现的错误；    
你还将学习模块json，它让你能够保存用户数据，以免在程序停止运行后丢失。     

学习处理文件和保存数据可让你的程序使用起来更容易：用户将能够选择输入什么样的数据，以及在什么时候输入；用户使用你的程序做一些工作后，可将程序关闭，以后再接着往下做。学习处理异常可帮助你应对文件不存在的情形，以及处理其他可能导致程序崩溃的问题。这让你的程序在面对错误的数据时更健壮——不管这些错误数据源自无意的错误，还是源自破坏程序的恶意企图。你在本章学习的技能可提高程序的适用性、可用性和稳定性。

# 10.1 从文件中读取数据

文本文件可存储的数据量多得难以置信：天气数据、交通数据、社会经济数据、文学作品等。每当需要分析或修改存储在文件中的信息时，读取文件都很有用，对数据分析应用程序来说尤其如此。例如，你可以编写一个这样的程序：读取一个文本文件的内容，重新设置这些数据的格式并将其写入文件，让浏览器能够显示这些内容。     

要使用文本文件中的信息，首先需要将信息读取到内存中。为此，你可以一次性读取文件的全部内容，也可以以每次一行的方式逐步读取。

## 10.1.1 读取整个文件

要读取文件，需要一个包含几行文本的文件。下面首先来创建一个文件，它包含精确到小数点后30位的圆周率值，且在小数点后每10位处都换行。

```
with open('pi_digits.txt') as file_object:
    contents = file_object.read()
    print(contents)
```

要以任何方式使用文件，都得**先打开文件，这样才能访问它**。函数open()接受一个参数：要打开的文件的名称。Python**在当前执行的文件所在的目录中查找指定的文件**。在这个示例中，当前运行的是file_reader.py，因此Python在file_reader.py所在的目录中查找pi_digits.txt。函数open()返回一个表示文件的对象。在这里，**`open('pi_digits.txt')`返回一个表示文件pi_digits.txt的对象；Python将这个对象存储在我们将在后面使用的变量中**。

**关键字with在不再需要访问文件后将其关闭**。在这个程序中，注意到我们调用了open()，但没有调用close()；你也可以调用open()和close()来打开和关闭文件，但这样做时，如果程序存在bug，导致close()语句未执行，文件将不会关闭。这看似微不足道，但未妥善地关闭文件可能会导致数据丢失或受损。如果在程序中过早地调用close()，你会发现需要使用文件时它已关闭（无法访问），这会导致更多的错误。并非在任何情况下都能轻松确定关闭文件的恰当时机，但通过使用前面所示的结构，可**让Python去确定：你只管打开文件，并在需要时使用它，Python自会在合适的时候自动将其关闭**。

有了表示pi_digits.txt的文件对象后，我们**使用方法read()（前述程序的第2行）读取这个文件的全部内容，并将其作为一个长长的字符串存储在变量contents中**。这样，通过打印contents的值，就可将这个文本文件的全部内容显示出来。

为何会多出这个空行呢？因为read()到达文件末尾时返回一个空字符串，而将这个空字符串显示出来时就是一个空行。要删除末尾的空行，可在print语句中使用rstrip()：
```
with open('pi_digits.txt') as file_object:
    contents = file_object.read()
    print(contents.rstrip())
```

## 10.1.2 文件路径

要让Python打开不与程序文件位于同一个目录中的文件，需要提供**文件路径**，它让Python到系统的特定位置去查找。     

由于文件夹text_files位于文件夹python_work中，因此可使用**相对文件路径**来打开该文件夹中的文件。相对文件路径让Python到指定的位置去查找，而该位置是相对于当前运行的程序所在目录的。在Linux和OS X中，你可以这样编写代码：
```
with open('text_files/filename.txt') as file_object:
```
在Windows系统中，在文件路径中使用反斜杠（\）而不是斜杠（/）：
```
with open('text_files\filename.txt') as file_object:
```
你还可以将文件在计算机中的**准确位置**告诉Python，这样就不用关心当前运行的程序存储在什么地方了。这称为**绝对文件路径**。在相对路径行不通时，可使用绝对路径。    

绝对路径通常比相对路径更长，因此将其存储在一个变量中，再将该变量传递给open()会有所帮助。在Linux和OS X中，绝对路径类似于下面这样：
```
file_path = '/home/ehmatthes/other_files/text_files/filename.txt'
with open(file_path) as file_object:
```

而在Windows系统中，它们类似于下面这样：   
```
file_path = 'C:\Users\ehmatthes\other_files\text_files\filename.txt'
with open(file_path) as file_object:
```

最简单的做法是，要么将数据文件存储在程序文件所在的目录，要么将其存储在程序文件所在目录下的一个文件夹（如text_files）中。

***注意***     

Windows系统有时能够正确地解读文件路径中的斜杠。如果你使用的是Windows系统，且结果不符合预期，请确保在文件路径中使用的是反斜杠。另外，由于反斜杠在Python中被视为转义标记，为在Windows中确保万无一失，应以原始字符串的方式指定路径，即在开头的单引号前加上r。**r'    '**

## 10.1.3 逐行读取

要以每次一行的方式检查文件，可对文件对象使用for循环：
```
filename = 'pi_digits.txt' 
with open(filename) as file_object: 
    for line in file_object: 
    print(line)
```
我们将要读取的文件的名称存储在变量filename中，这是使用文件时一种常见的做法。由于变量filename表示的并非实际文件——它只是一个让Python**知道到哪里去查找文件的字符串**，因此可轻松地将'pi_digits.txt'替换为你要使用的另一个文件的名称。调用open()后，将一个表示文件及其内容的对象存储到了变量file_object中。这里也使用了关键字with，让Python负责妥善地打开和关闭文件。为查看文件的内容，我们通过对文件对象执行循环来遍历文件中的每一行。我们打印每一行时，发现空白行更多了。     

为何会出现这些空白行呢？因为在这个文件中，**每行的末尾都有一个看不见的换行符，而print语句也会加上一个换行符，因此每行末尾都有两个换行符：一个来自文件，另一个来自print语句**。要**消除这些多余的空白行，可在print语句中使用rstrip()**。

## 10.1.4 创建一个包含文件各行内容的列表

使用关键字with时，open()返回的文件对象只在with代码块内可用。如果要在with代码块外访问文件的内容，可在with代码块内将文件的各行存储在一个列表中，并在with代码块外使用该列表：你可以立即处理文件的各个部分，也可推迟到程序后面再处理。下面的示例在with代码块中将文件pi_digits.txt的各行存储在一个列表中，再在with代码块外打印它们：
```
filename = 'pi_digits.txt'
with open(filename) as file_object:
    lines = file_object.readlines()
for line in lines:
    print(line.rstrip())
```
**方法readlines()从文件中读取每一行，并将其存储在一个列表中**；接下来，该**列表被存储到变量lines中；在with代码块外，我们依然可以使用这个变量**。接着使用一个简单的for循环来打印lines中的各行。由于列表lines的每个元素都对应于文件中的一行，因此输出与文件内容完全一致。

## 10.1.5 使用文件的内容

将文件读取到内存中后，就可以以任何方式使用这些数据了。下面以简单的方式使用圆周率的值。首先，我们将创建一个字符串，它包含文件中存储的所有数字，且没有任何空格：
```
filename = 'pi_digits.txt'
with open(filename) as file_object:
    lines = file_object.readlines()
pi_string = ''
for line in lines:
    pi_string+= line.rstrip()
print(pi_string)
print(len(pi_string))
```
创建了一个变量——pi_string，用于存储圆周率的值。接下来，我们使用一个循环将各行都加入pi_string，并删除每行末尾的换行符。我们打印这个字符串及其长度。在变量pi_string存储的字符串中，包含原来位于每行左边的空格，**为删除这些空格，可使用strip()**而不是rstrip()：
`pi_string+= line.strip()`
这样，我们就获得了一个这样的字符串：它包含精确到30位小数的圆周率值。这个字符串长32字符，因为它还包含整数部分的3和小数点。

***注意***    

读取文本文件时，Python将其中的所有文本都解读为字符串。如果你读取的是数字，并要将其作为数值使用，就**必须使用函数int()将其转换为整数，或使用函数float()将其转换为浮点数**。

## 10.1.6 包含一百万位的大型文件

在这里，我们只打印到小数点后50位，以免终端为显示全部1000000位而不断地翻滚：
```
filename = 'pi_million_digits.txt'
with open(filename) as file_object:
    lines = file_object.readlines()
pi_string = ''
for line in lines:
    pi_string+= line.rstrip()
print(pi_string[:52]+"...")
print(len(pi_string))
```
输出表明，我们创建的字符串确实包含精确到小数点后1000000位的圆周率值。对于你可处理的数据量，Python没有任何限制；只要系统的内存足够多，你想处理多少数据都可以。

## 10.1.7 圆周率值中包含你的生日吗

以确定某个人的生日是否包含在圆周率值的前1000000位中。为此，可将生日表示为一个由数字组成的字符串，再检查这个字符串是否包含在pi_string中：
```
filename = 'pi_million_digits.txt'

with open(filename) as file_object:
    lines = file_object.readlines()

pi_string = ''
for line in lines:
    pi_string += line.strip()

birthday = input("Enter your birthday, in the form mmddyy: ")
if birthday in pi_string:
    print("Your birthday appears in the first million digits of pi!")
else:
    print("Your birthday does not appear in the first million digits of pi.")
```
我们提示用户输入其生日，接下来，我们检查这个字符串是否包含在pi_string中。

## 动手试一试

### 10-1 Python学习笔记：      
在文本编辑器中新建一个文件，写几句话来总结一下你至此学到的Python知识，其中每一行都以“In Python you can”打头。将这个文件命名为learning_python.txt，并将其存储到为完成本章练习而编写的程序所在的目录中。编写一个程序，它读取这个文件，并将你所写的内容打印三次：第一次打印时读取整个文件；第二次打印时遍历文件对象；第三次打印时将各行存储在一个列表中，再在with代码块外打印它们。

```
filename = 'learning_python.txt'
with open (filename) as file_object:
    # 读取整个文件
    contents = file_object.read()
    print(contents)

with open(filename) as file_object:
    # 遍历文件对象
    for line in file_object:
        print(line.rstrip())

with open(filename) as file_object:
    # 将各行存储在一个列表中，在with代码块外打印它们
    lines = file_object.readlines()
for line in lines:
    print(line.rstrip())

```

### 10-2 C语言学习笔记：       
可使用**方法replace()将字符串中的特定单词都替换为另一个单词**。下面是一个简单的示例，演示了如何将句子中的'dog'替换为'cat'：
```
>>> message = "I really like dogs."
>>> message.replace('dog', 'cat')
'I really like cats.'
```
读取你刚创建的文件learning_python.txt中的每一行，将其中的Python都替换为另一门语言的名称，如C。将修改后的各行都打印到屏幕上。

```
filename = 'learning_python.txt'
with open (filename) as file_object:
    for line in file_object:
        newline = line.replace('Python', 'C')
        print(newline.rstrip())

```
replace并非永久替换，因此要将替换后的结果赋给一个新变量。

# 10.2 写入文件

保存数据的最简单的方式之一是将其写入到文件中。通过将输出写入文件，即便关闭包含程序输出的终端窗口，这些输出也依然存在：你可以在程序结束运行后查看这些输出，可与别人分享输出文件，还可编写程序来将这些输出读取到内存中并进行处理。

## 10.2.1 写入空文件

要将文本写入文件，你在调用open()时需要提供另一个实参，告诉Python你要写入打开的文件。为明白其中的工作原理，我们来将一条简单的消息存储到文件中，而不是将其打印到屏幕上：
```
filename = 'programming.txt'
with open(filename, 'w') as file_object: 
    file_object.write("I love programming.") 
```
**调用open()时提供了两个实参**。第一个实参也是要打开的文件的名称；第二个实参（'w'）告诉Python，我们要以写入模式打开这个文件。打开文件时，**可指定读取模式（'r'）、写入模式（'w'）、附加模式（'a'）或让你能够读取和写入文件的模式（'r+'）**。如果你省略了模式实参，Python将**以默认的只读模式打开文件**。    

如果你要写入的文件不存在，函数open()将自动创建它。然而，以写入（'w'）模式打开文件时千万要小心，因为如果指定的文件已经存在，Python将在返回文件对象前清空该文件。以'w'模式写入文件时，如果文件已存在，会直接覆盖（相当于删掉后新写入一个文件）。如果我们希望追加到文件末尾怎么办？可以传入'a'以追加（append）模式写入。       

我们使用文件对象的**方法write()**将一个字符串写入文件。这个程序没有终端输出，但如果你打开文件programming.txt，将看到其中包含的内容。

***注意***     
Python只能将字符串写入文本文件。要将数值数据存储到文本文件中，**必须先使用函数str()将其转换为字符串格式**。

## 10.2.2 写入多行

**函数write()不会在你写入的文本末尾添加换行符**，因此如果你写入多行时没有指定换行符，如果你打开programming.txt，将发现两行内容挤在一起。    
要让每个字符串都单独占一行，需要在write()语句中包含换行符：
```
filename = 'programming.txt'
with open(filename, 'w') as file_object:
file_object.write("I love programming.\n")
file_object.write("I love creating new games.\n")
```
像显示到终端的输出一样，还可以使用空格、制表符和空行来设置这些输出的格式。

## 10.2.3 附加到文件

如果你要给文件添加内容，而不是覆盖原有的内容，可以附加模式打开文件。你以附加模式打开文件时，Python不会在返回文件对象前清空文件，而你写入到文件的行都将添加到文件末尾。如果指定的文件不存在，Python将为你创建一个空文件。
```
filename = 'programming.txt'

with open(filename, 'a') as file_object:
    file_object.write("I also love finding meaning in large datasets.\n")
    file_object.write("I love creating apps that can run in a browser.\n")

```
我们打开文件时指定了实参'a'，以便将内容附加到文件末尾，而不是覆盖文件原来的内容。我们又写入了两行，它们被添加到文件programming.txt末尾。

## 动手试一试

### 10-3 访客：        
编写一个程序，提示用户输入其名字；用户作出响应后，将其名字写入到文件guest.txt中。

```
name = input("Please type your name here:")
filename = 'guest.txt'
with open (filename, 'w') as file_object:
    file_object.write(name.title())

```

### 10-4 访客名单：        
编写一个while循环，提示用户输入其名字。用户输入其名字后，在屏幕上打印一句问候语，并将一条访问记录添加到文件guest_book.txt中。确保这个文件中的每条记录都独占一行。

```
filename = 'guest_book.txt'

print("Enter 'quit' when you are finished")
while True:
    name = input("\nWhat's your name?")
    if name == 'quit':
        break
    else:
        with open(filename, 'a') as f:
            f.write(name+"\n")
        print("Hi "+name+", you've been added to the guest book.")
```

### 10-5 关于编程的调查：       
编写一个while循环，询问用户为何喜欢编程。每当用户输入一个原因后，都将其添加到一个存储所有原因的文件中。

```
filename = 'programming_poll.txt'

responses = []
while True:
	response = input("\nWhy do you like programming? ")
	responses.append(response)

	continue_poll = input("Would you like to let someone else respond? (y/n) ")
	if continue_poll != 'y':
		break

with open(filename, 'a') as f:
	for response in responses:
		f.write(response+"\n")
```

# 10.3 异常

Python使用被称为异常的特殊对象来管理程序执行期间发生的错误。每当发生让Python不知所措的错误时，它都会创建一个异常对象。如果你编写了处理该异常的代码，程序将继续运行；如果你未对异常进行处理，程序将停止，并显示一个traceback，其中包含有关异常的报告。     

异常是使用try-except代码块处理的。try-except代码块让Python执行指定的操作，同时告诉Python发生异常时怎么办。使用了try-except代码块时，即便出现异常，程序也将继续运行：显示你编写的友好的错误消息，而不是令用户迷惑的traceback。

## 10.3.1 处理ZeroDivisionError异常

下面来看一种导致Python引发异常的简单错误。你可能知道不能将一个数字除以0，但我们还是让Python这样做吧：


```python
print(5/0)
```


    ---------------------------------------------------------------------------

    ZeroDivisionError                         Traceback (most recent call last)

    <ipython-input-2-fad870a50e27> in <module>
    ----> 1 print(5/0)
    

    ZeroDivisionError: division by zero


在上述traceback中，指出的错误ZeroDivisionError是一个异常对象。Python无法按你的要求做时，就会创建这种对象。在这种情况下，Python将停止运行程序，并指出引发了哪种异常，而我们可根据这些信息对程序进行修改。下面我们将告诉Python，发生这种错误时怎么办；这样，如果再次发生这样的错误，我们就有备无患了。

## 10.3.2 使用try-except代码块

当你认为可能发生了错误时，可编写一个try-except代码块来处理可能引发的异常。你让Python尝试运行一些代码，并告诉它如果这些代码引发了指定的异常，该怎么办。       

处理ZeroDivisionError异常的try-except代码块类似于下面这样：


```python
try:
    print(5/0)
except ZeroDivisionError:
    print("You can't divide by zero!")
```

    You can't divide by zero!


我们将导致错误的代码行print(5/0)放在了一个try代码块中。如果try代码块中的代码运行起来没有问题，Python将跳过except代码块；如果try代码块中的代码导致了错误，Python将查找这样的except代码块，并运行其中的代码，即其中指定的错误与引发的错误相同。     

在这个示例中，try代码块中的代码引发了ZeroDivisionError异常，因此Python指出了该如何解决问题的except代码块，并运行其中的代码。这样，用户看到的是一条友好的错误消息，而不是traceback。   

如果try-except代码块后面还有其他代码，程序将接着运行，因为已经告诉了Python如何处理这种错误。下面来看一个捕获错误后程序将继续运行的示例。

## 10.3.3 使用异常避免崩溃

发生错误时，如果程序还有工作没有完成，妥善地处理错误就尤其重要。这种情况经常会出现在要求用户提供输入的程序中；如果程序能够妥善地处理无效输入，就能再提示用户提供有效输入，而不至于崩溃。       

程序崩溃可不好，但让用户看到traceback也不是好主意。不懂技术的用户会被它们搞糊涂，而且如果用户怀有恶意，他会通过traceback获悉你不希望他知道的信息。例如，他将知道你的程序文件的名称，还将看到部分不能正确运行的代码。有时候，训练有素的攻击者可根据这些信息判断出可对你的代码发起什么样的攻击。

## 10.3.4 else代码块

通过将可能引发错误的代码放在try-except代码块中，可提高这个程序抵御错误的能力。错误是执行除法运算的代码行导致的，因此我们需要将它放到try-except代码块中。这个示例还包含一个else代码块；依赖于try代码块成功执行的代码都应放到else代码块中


```python
print("Give me two numbers, and I'll divide them.")
print("Enter 'q' to quit.")

while True:
    first_number = input("\nFirst number: ")
    if first_number == 'q':
        break
    second_number = input("Second number: ")
    try:
        answer = int(first_number) / int(second_number)
    except ZeroDivisionError:
        print("You can't divide by 0!")
    else:
        print(answer)
```

    Give me two numbers, and I'll divide them.
    Enter 'q' to quit.
    
    First number: 3
    Second number: 1
    3.0
    
    First number: 3
    Second number: 2
    1.5
    
    First number: 5
    Second number: 8
    0.625
    
    First number: 3
    Second number: 0
    You can't divide by 0!
    
    First number: q


我们让Python尝试执行try代码块中的除法运算，这个代码块只包含可能导致错误的代码。依赖于try代码块成功执行的代码都放在else代码块中；在这个示例中，如果除法运算成功，我们就使用else代码块来打印结果。如果try代码块因除零错误而失败，我们就打印一条友好的消息，告诉用户如何避免这种错误。程序将继续运行，用户根本看不到traceback。    

try-except-else代码块的工作原理大致如下：Python尝试执行try代码块中的代码；只有可能引发异常的代码才需要放在try语句中。有时候，有一些仅在try代码块成功执行时才需要运行的代码；这些代码应放在else代码块中。except代码块告诉Python，如果它尝试运行try代码块中的代码时引发了指定的异常，该怎么办。


## 10.3.5 处理FileNotFoundError异常

使用文件时，一种常见的问题是找不到文件：你要查找的文件可能在其他地方、文件名可能不正确或者这个文件根本就不存在。对于所有这些情形，都可使用try-except代码块以直观的方式进行处理。


```python
filename = 'alice.txt'

try:
    with open(filename) as f_obj:
        contents = f_obj.read()
except FileNotFoundError as e:
    msg = "Sorry, the file " + filename + " does not exist."
    print(msg)
else:
    # Count the approximate number of words in the file.
    words = contents.split()
    num_words = len(words)
    print("The file " + filename + " has about " + str(num_words) + " words.")
```

    Sorry, the file alice.txt does not exist.


在这个示例中，try代码块引发FileNotFoundError异常，因此Python找出与该错误匹配的except代码块，并运行其中的代码。最终的结果是显示一条友好的错误消息，而不是traceback。    

如果文件不存在，这个程序什么都不做，因此错误处理代码的意义不大。下面来扩展这个示例，看看在你使用多个文件时，异常处理可提供什么样的帮助。

## 10.3.6 分析文本

下面来提取童话Alice in Wonderland的文本，并尝试计算它包含多少个单词。我们将使用方法split()，它根据一个字符串创建一个单词列表。下面是对只包含童话名"Alice in Wonderland"的字符串调用方法split()的结果：


```python
title = 'Alice in Wonderland'
print(title.split())
```

    ['Alice', 'in', 'Wonderland']


**方法split()以空格为分隔符将字符串分拆成多个部分，并将这些部分都存储到一个列表中**。结果是一个包含字符串中所有单词的列表，虽然有些单词可能包含标点。为计算Alice in Wonderland包含多少个单词，我们将对整篇小说调用split()，再计算得到的列表包含多少个元素，从而确定整篇童话大致包含多少个单词。

```
filename = 'alice.txt'

try:
    with open(filename) as f_obj:
        contents = f_obj.read()
except FileNotFoundError as e:
    msg = "Sorry, the file " + filename + " does not exist."
    print(msg)
else:
    # Count the approximate number of words in the file.
    words = contents.split()
    num_words = len(words)
    print("The file " + filename + " has about " + str(num_words) + " words.")
```

我们把文件alice.txt移到了正确的目录中，让try代码块能够成功地执行。我们对变量contents（它现在是一个长长的字符串，包含童话Alice in Wonderland的全部文本）调用方法split()，以生成一个列表，其中包含这部童话中的所有单词。当我们使用len()来确定这个列表的长度时，就知道了原始字符串大致包含多少个单词。接着，我们打印一条消息，指出文件包含多少个单词。这些代码都放在else代码块中，因为仅当try代码块成功执行时才执行它们。输出指出了文件alice.txt包含多少个单词。

## 10.3.7 使用多个文件

下面多分析几本书。这样做之前，我们先将这个程序的大部分代码移到一个名为count_words()的函数中，这样对多本书进行分析时将更容易。    

现在可以编写一个简单的循环，计算要分析的任何文本包含多少个单词了。为此，我们将要分析的文件的名称存储在一个列表中，然后对列表中的每个文件都调用count_words()。我们将尝试计算Alice in Wonderland、Siddhartha、Moby Dick和Little Women分别包含多少个单词。    

使用try-except代码块提供了两个重要的优点：避免让用户看到traceback；让程序能够继续分析能够找到的其他文件。如果不捕获因找不到siddhartha.txt而引发的FileNotFoundError异常，用户将看到完整的traceback，而程序将在尝试分析Siddhartha后停止运行——根本不分析Moby Dick和Little Women。

## 10.3.8 失败时一声不吭

有时候你希望程序在发生异常时一声不吭，就像什么都没有发生一样继续运行。Python有一个pass语句，可在代码块中使用它来让Python什么都不要做：

```
def count_words(filename):
    """Count the approximate number of words in a file."""
    try:
        with open(filename) as f_obj:
            contents = f_obj.read() 
    except FileNotFoundError:
        pass
    else:
        # Count approximate number of words in the file.
        words = contents.split()
        num_words = len(words)
        print("The file " + filename + " has about " + str(num_words) + " words.")

filenames = ['alice.txt', 'siddhartha.txt', 'moby_dick.txt', 'little_women.txt']
for filename in filenames:
    count_words(filename)
```

现在，出现FileNotFoundError异常时，将执行except代码块中的代码，但什么都不会发生。这种错误发生时，不会出现traceback，也没有任何输出。用户将看到存在的每个文件包含多少个单词，但没有任何迹象表明有一个文件未找到。    

pass语句还充当了占位符，它提醒你在程序的某个地方什么都没有做，并且以后也许要在这里做些什么。例如，在这个程序中，我们可能决定将找不到的文件的名称写入到文件missing_files.txt中。用户看不到这个文件，但我们可以读取这个文件，进而处理所有文件找不到的问题。

## 10.3.9 决定报告哪些错误

Python的错误处理结构让你能够细致地控制与用户分享错误信息的程度，要分享多少信息由你决定。     

编写得很好且经过详尽测试的代码不容易出现内部错误，如语法或逻辑错误，但只要程序依赖于外部因素，如用户输入、存在指定的文件、有网络链接，就有可能出现异常。凭借经验可判断该在程序的什么地方包含异常处理块，以及出现错误时该向用户提供多少相关的信息。

## 动手试一试

### 10-6 加法运算：      
提示用户提供数值输入时，常出现的一个问题是，用户提供的是文本而不是数字。在这种情况下，当你尝试将输入转换为整数时，将引发ValueError异常。编写一个程序，提示用户输入两个数字，再将它们相加并打印结果。在用户输入的任何一个值不是数字时都捕获ValueError异常，并打印一条友好的错误消息。对你编写的程序进行测试：先输入两个数字，再输入一些文本而不是数字。


```python
try:
    x = input("Give me a number: ")
    x = int(x)
    
    y = input("Give me another number: ")
    y = int(y)    
except ValueError:
    print("Sorry, I really needed a number.")
else:
    sum = x + y
    print("The sum of "+str(x)+" and "+str(y)+" is "+str(sum)+".")
```

    Give me a number: fred
    Sorry, I really needed a number.


### 10-7 加法计算器：      
将你为完成练习10-6而编写的代码放在一个while循环中，让用户犯错（输入的是文本而不是数字）后能够继续输入数字。


```python
print("Enter 'q' at any time to quit.\n")

while True:
    try:
        x = input("\nGive me a number: ")
        if x == 'q':
            break
        x = int(x)
        
        y = input("Give me another number: ")
        if y == 'q':
            break
        y = int(y)
    except ValueError:
        print("Sorry, I really need a number.")
    else:
        sum = x + y
        print("The sum of "+str(x)+" and "+str(y)+" is "+str(sum)+".")
```

    Enter 'q' at any time to quit.
    
    
    Give me a number: 23
    Give me another number: 6
    The sum of 23 and 6 is 29.
    
    Give me a number: three
    Sorry, I really need a number.
    
    Give me a number: 12
    Give me another number: 20
    The sum of 12 and 20 is 32.
    
    Give me a number: q


### 10-8 猫和狗：       
创建两个文件cats.txt和dogs.txt，在第一个文件中至少存储三只猫的名字，在第二个文件中至少存储三条狗的名字。编写一个程序，尝试读取这些文件，并将其内容打印到屏幕上。将这些代码放在一个try-except代码块中，以便在文件不存在时捕获FileNotFound错误，并打印一条友好的消息。将其中一个文件移到另一个地方，并确认except代码块中的代码将正确地执行。

```
filenames = ['cats.txt', 'dogs.txt']

for filename in filenames:
	print("\nReading file: "+filename)
	
	try:
		with open(filename) as f:
			contents = f.read()
			print(contents)
	except FileNotFoundError:
		print("Sorry, I can't find that file.")
```

### 10-9 沉默的猫和狗：      
修改你在练习10-8中编写的except代码块，让程序在文件不存在时一言不发。

```
filenames = ['cats.txt', 'dogs.txt']

for filename in filenames:
	try:
		with open(filename) as f:
			contents = f.read()
	except FileNotFoundError:
		pass
	else:
		print("\nReading file: "+filename)
		print(contents)
```

### 10-10 常见单词：         
访问项目Gutenberg（http://gutenberg.org/)
并找一些你想分析的图书。下载这些作品的文本文件或将浏览器中的原始文本复制到文本文件中。       

你可以使用方法count()来确定特定的单词或短语在字符串中出现了多少次。例如，下面的代码计算'row'在一个字符串中出现了多少次：
```
>>line = "Row, row, row your boat"
>>line.count('row')
2
>>line.lower().count('row')
3
```
请注意，通过使用lower()将字符串转换为小写，可捕捉要查找的单词出现的所有次数，而不管其大小写格式如何。    

编写一个程序，它读取你在项目Gutenberg中获取的文件，并计算单词'the'在每个文件中分别出现了多少次

```
filenames = ['a_tale_of_two_cities.txt', 'pride_and_prejudice.txt', 'learning_python.txt']

for filename in filenames:
	try:
		with open(filename) as f:
			contents = f.read()
	except FileNotFoundError:
		pass
	else:
		print("\nReading file: "+filename)
		print("The number of'the'appears in the file:")
		print(contents.lower().count('the'))
```

# 10.4 存储数据

很多程序都要求用户输入某种信息，如让用户存储游戏首选项或提供要可视化的数据。不管专注的是什么，程序都把用户提供的信息存储在列表和字典等数据结构中。用户关闭程序时，你几乎总是要保存他们提供的信息；一种简单的方式是使用模块json来存储数据。      

模块json让你能够将简单的Python数据结构转储到文件中，并在程序再次运行时加载该文件中的数据。你还可以使用json在Python程序之间分享数据。更重要的是，JSON数据格式并非Python专用的，这让你能够将以JSON格式存储的数据与使用其他编程语言的人分享。这是一种轻便格式，很有用，也易于学习。     

***注意***      
JSON（JavaScript Object Notation）格式最初是为JavaScript开发的，但随后成了一种常见格式，被包括Python在内的众多语言采用。

## 10.4.1 使用json.dump()和json.load()

我们来编写一个存储一组数字的简短程序，再编写一个将这些数字读取到内存中的程序。第一个程序将使用json.dump()来存储这组数字，而第二个程序将使用json.load()。    

**函数json.dump()接受两个实参：要存储的数据以及可用于存储数据的文件对象**。下面演示了如何使用json.dump()来存储数字列表：
```
import json

numbers = [2, 3, 5, 7, 11, 13]

filename = 'numbers.json'
with open(filename, 'w') as file_object:
    json.dump(numbers, file_object)
```
我们先导入模块json，再创建一个数字列表。我们指定了要将该数字列表存储到其中的文件的名称。通常使用文件扩展名.json来指出文件存储的数据为JSON格式。接下来，我们以写入模式打开这个文件，让json能够将数据写入其中。我们使用函数json.dump()将数字列表存储到文件numbers.json中。

下面再编写一个程序，使用json.load()将这个列表读取到内存中：
```
import json

filename = 'numbers.json'
with open(filename) as file_object:
    numbers = json.load(file_object)
print(numbers)
```
我们确保读取的是前面写入的文件。这次我们以读取方式打开这个文件，因为Python只需读取这个文件。我们使用函数json.load()加载存储在numbers.json中的信息，并将其存储到变量numbers中。最后，我们打印恢复的数字列表，看看它是否与number_writer.py中创建的数字列表相同。这是一种在程序之间共享数据的简单方式。

## 10.4.2 保存和读取用户生成的数据

对于用户生成的数据，使用json保存它们大有裨益，因为如果不以某种方式进行存储，等程序停止运行时用户的信息将丢失。下面来看一个这样的例子：用户首次运行程序时被提示输入自己的名字，这样再次运行程序时就记住他了。
```
import json
username = input("What is your name? ") 
filename = 'username.json'
with open(filename, 'w') as f_obj:
    json.dump(username, f_obj) 
    print("We'll remember you when you come back, "+username+"!")
```

现在再编写一个程序，向其名字被存储的用户发出问候：
```
import json
filename = 'username.json'
with open(filename) as f_obj:
    username = json.load(f_obj)
    print("Welcome back, "+username+"!") 
```
我们使用json.load()将存储在username.json中的信息读取到变量username中。恢复用户名后，我们就可以欢迎用户回来了。

我们需要将这两个程序合并到一个程序（remember_me.py）中。这个程序运行时，我们将尝试从文件username.json中获取用户名，因此我们首先编写一个尝试恢复用户名的try代码块。如果这个文件不存在，我们就在except代码块中提示用户输入用户名，并将其存储在username.json中，以便程序再次运行时能够获取它：    
```
import json
# 如果以前存储了用户名，就加载它
#  否则，就提示用户输入用户名并存储它
filename = 'username.json'
try:
    with open(filename) as f_obj:
        username = json.load(f_obj) 
except FileNotFoundError:
    username = input("What is your name? ")
    with open(filename, 'w') as f_obj:
        json.dump(username, f_obj)
        print("We'll remember you when you come back, "+username+"!")
else:
    print("Welcome back, "+username+"!")
```


## 10.4.3 重构

将代码划分为一系列完成具体工作的函数。这样的过程被称为**重构**。重构让代码更清晰、更易于理解、更容易扩展。    

要重构remember_me.py，可将其大部分逻辑放到一个或多个函数中。remember_me.py的重点是问候用户，因此我们将其所有代码都放到一个名为greet_user()的函数中。这个程序更清晰些，但函数greet_user()所做的不仅仅是问候用户，还在存储了用户名时获取它，而在没有存储用户名时提示用户输入一个。     

下面来重构greet_user()，让它不执行这么多任务。为此，我们首先将获取存储的用户名的代码移到另一个函数中。    

新增的函数get_stored_username()目标明确。如果存储了用户名，这个函数就获取并返回它；如果文件username.json不存在，这个函数就返回None。这是一种不错的做法：函数要么返回预期的值，要么返回None；这让我们能够使用函数的返回值做简单测试。如果成功地获取了用户名，就打印一条欢迎用户回来的消息，否则就提示用户输入用户名。     

我们还需将greet_user()中的另一个代码块提取出来：将没有存储用户名时提示用户输入的代码放在一个独立的函数中：

```
import json

def get_stored_username():
    """Get stored username if available."""
    filename = 'username.json'
    try:
        with open(filename) as f_obj:
            username = json.load(f_obj)
    except FileNotFoundError:
        return None
    else:
        return username

def get_new_username():
    """Prompt for a new username."""
    username = input("What is your name? ")
    filename = 'username.json'
    with open(filename, 'w') as f_obj:
        json.dump(username, f_obj)
    return username

def greet_user():
    """Greet the user by name."""
    username = get_stored_username()
    if username:
        print("Welcome back, " + username + "!")
    else:
        username = get_new_username()
        print("We'll remember you when you come back, " + username + "!")

greet_user()
```

在remember_me.py的这个最终版本中，每个函数都执行单一而清晰的任务。我们调用greet_user()，它打印一条合适的消息：要么欢迎老用户回来，要么问候新用户。为此，它首先调用get_stored_username()，这个函数只负责获取存储的用户名（如果存储了的话），再在必要时调用get_new_username()，这个函数只负责获取并存储新用户的用户名。要编写出清晰而易于维护和扩展的代码，这种划分工作必不可少。

## 动手试一试

### 10-11 喜欢的数字：        
编写一个程序，提示用户输入他喜欢的数字，并使用json.dump()将这个数字存储到文件中。再编写一个程序，从文件中读取这个值，并打印消息“I know your favorite number! It's _____.”。

```
import json

number = input("What's your favorite number? ")

with open('favorite_number.json', 'w') as f:
	json.dump(number, f)
	print("Thanks! I'll remember that.")
```

```
import json

with open('favorite_number.json') as f:
	number = json.load(f)
print("I know your favorite number! It's "+str(number)+".")
```

### 10-12 记住喜欢的数字：       
将练习10-11中的两个程序合而为一。如果存储了用户喜欢的数字，就向用户显示它，否则提示用户输入他喜欢的数字并将其存储到文件中。运行这个程序两次，看看它是否像预期的那样工作。

```
import json
try:
	with open('favorite_number.json') as f:
		number = json.load(f)
except FileNotFoundError:
	number = input("What's your favorite number? ")
	with open('favorite_number.json', 'w') as f:
		json.dump(number, f)
	print("Thanks, I'll remember that.")
else:
	print("I know your favorite number! It's "+str(number)+".")
```

### 10-13 验证用户：          
最后一个remember_me.py版本假设用户要么已输入其用户名，要么是首次运行该程序。我们应修改这个程序，以应对这样的情形：当前和最后一次运行该程序的用户并非同一个人。      

为此，在greet_user()中打印欢迎用户回来的消息前，先询问他用户名是否是对的。如果不对，就调用get_new_username()让用户输入正确的用户名。

```
import json

def get_stored_username():
    """Get stored username if available."""
    filename = 'username.json'
    try:
        with open(filename) as f_obj:
            username = json.load(f_obj)
    except FileNotFoundError:
        return None
    else:
        return username

def get_new_username():
    """Prompt for a new username."""
    username = input("What is your name? ")
    filename = 'username.json'
    with open(filename, 'w') as f_obj:
        json.dump(username, f_obj)
    return username

def greet_user():
    """Greet the user by name."""
    username = get_stored_username()
    if username:
        correct = input("Are you " + username + "? (y/n) ")
        if correct == 'y':
            print("Welcome back, " + username + "!")
        else:
            username = get_new_username()
            print("We'll remember you when you come back, " + username + "!")
    else:
        username = get_new_username()
        print("We'll remember you when you come back, " + username + "!")

greet_user()
```

第二种更加简洁的方式：
```
import json

def get_stored_username():
    """Get stored username if available."""
    filename = 'username.json'
    try:
        with open(filename) as f_obj:
            username = json.load(f_obj)
    except FileNotFoundError:
        return None
    else:
        return username

def get_new_username():
    """Prompt for a new username."""
    username = input("What is your name? ")
    filename = 'username.json'
    with open(filename, 'w') as f_obj:
        json.dump(username, f_obj)
    return username

def greet_user():
    """Greet the user by name."""
    username = get_stored_username()
    if username:
        correct = input("Are you " + username + "? (y/n) ")
        if correct == 'y':
            print("Welcome back, " + username + "!")
            return

    username = get_new_username()
    print("We'll remember you when you come back, " + username + "!")

greet_user()
```

One way to clean this function up is to use an empty `return` statement. An empty `return` statement tells Python to leave the function without running any more code in the function.    

The `return` statement means the code in the function stops running after printing the welcome back message. When the username doesn’t exist, or the username is incorrect, the `return` statement is never reached. The second part of the function will only run when the `if` statements fail, so we don’t need an `else` block. Now the function prompts for a new username when either `if` statement fails.

# 10.5 小结

在本章中，学习了：     
1. 如何使用文件；     
2. 如何一次性读取整个文件，以及如何以每次一行的方式读取文件的内容；      
3. 如何写入文件，以及如何将文本附加到文件末尾；    
4. 什么是异常以及如何处理程序可能引发的异常；     
5. 如何存储Python数据结构，以保存用户提供的信息，避免用户每次运行程序时都需要重新提供。
